### PR TITLE
Metadata rendering

### PR DESCRIPTION
**Pull request by Alexey-ebi**

### CODE REVIEW

File: src/app/ruleset-experiment/ruleset-experiment.component.html [hunk 1]
```diff
@@ -33,7 +33,98 @@ <h2>FAANG Rule sets
...
+            <td>{{rule.description}}</td>
+            <td>{{rule.type}}</td>
+            <td>{{rule.mandatory}}</td>
+            <td>{{allowMultiple(rule)}}</td>
+            <td>{{getValidItems(rule, 'valid_values')}}</td>
+            <td>{{getValidItems(rule, 'valid_units')}}</td>
+            <td>
+              <ng-container *ngIf="rule['valid_terms']">
+                <div *ngFor="let link of rule['valid_terms']">
+                  <a href="https://www.ebi.ac.uk/ols/ontologies/obi/terms?iri={{link['term_iri']}}" >{{getLink(link['term_iri'])}}</a>
```
Comment (wizardfan) [inline src/app/ruleset-experiment/ruleset-experiment.component.html]: use constants.ts ols_prefix instead

File: src/app/ruleset-analysis/ruleset-analysis.component.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ export class RulesetAnalysisComponent implements OnInit, AfterViewChecked {
   clicked = false;
   fragment: string;
 
-  constructor(private titleService: Title, private apiFileService: ApiFileService, private route: ActivatedRoute) { }
+  constructor(private titleService: Title, private apiFileService: ApiDataService, private route: ActivatedRoute) { }
```
Comment (wizardfan) [inline src/app/ruleset-analysis/ruleset-analysis.component.ts]: variable name needs chang accordingly

File: src/app/app-routing.module.ts [hunk 3]
```diff
@@ -22,6 +22,9 @@ import {DatasetsSummaryComponent} from './datasets-summary/datasets-summary.comp
 import {FilesSummaryComponent} from './files-summary/files-summary.component';
 import {ProtocolAnalysisComponent} from './protocol-analysis/protocol-analysis.component';
 import {NonExistingComponent} from './non-existing/non-existing.component';
+import {RulesetSampleComponent} from './ruleset-sample/ruleset-sample.component';
+import {RulesetExperimentComponent} from './ruleset-experiment/ruleset-experiment.component';
+import {RulesetAnalysisComponent} from './ruleset-analysis/ruleset-analysis.component';
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/app-routing.module.ts]: Done

[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/app-routing.module.ts]: Refactoring for summary and other components will be done in another branch and PR. It's possible to move other common functions to constants too

File: src/app/ruleset-experiment/ruleset-experiment.component.css [hunk 4]
```diff
@@ -0,0 +1,8 @@
+button {
+  margin-left: 5px;
+  font-size: 14px;
+}
+
+i {
+  margin-left: 5px;
+}
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-experiment/ruleset-experiment.component.css]: Done

File: src/app/ruleset-sample/ruleset-sample.component.ts [hunk 5]
```diff
@@ -1,17 +1,27 @@
...
 @Component({
   selector: 'app-ruleset-sample',
   templateUrl: './ruleset-sample.component.html',
   styleUrls: ['./ruleset-sample.component.css']
 })
 export class RulesetSampleComponent implements OnInit {
+  error: string;
 
-  constructor(private titleService: Title) { }
+  constructor(private titleService: Title, private apiFileService: ApiFileService) { }
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: Done

File: src/app/ruleset-analysis/ruleset-analysis.component.ts [hunk 6]
```diff
@@ -1,17 +1,29 @@
 import { Component, OnInit } from '@angular/core';
 import {Title} from '@angular/platform-browser';
+import {ApiFileService} from '../services/api-file.service';
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-analysis/ruleset-analysis.component.ts]: Done

File: src/app/ruleset-sample/ruleset-sample.component.ts [hunk 7]
```diff
@@ -113,4 +115,8 @@ export class RulesetSampleComponent implements OnInit, AfterViewChecked {
     }
   }
 
+  getLength(data: any) {
+    return data.length;
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: Done

File: src/app/ruleset-analysis/ruleset-analysis.component.ts (hunk 2 above)
Comment (Alexey-ebi) [reply src/app/ruleset-analysis/ruleset-analysis.component.ts]: Done

File: src/app/services/api-file.service.ts [hunk 8]
```diff
@@ -303,6 +303,17 @@ export class ApiFileService {
     );
   }
 
+  getRulesetSample() {
+    const url = 'https://raw.githubusercontent.com/FAANG/dcc-metadata/master/rulesets/faang_samples.metadata_rules.json';
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/services/api-file.service.ts]: Done

File: src/app/ruleset-sample/ruleset-sample.component.ts [hunk 9]
```diff
@@ -1,29 +1,49 @@
...
+
+  ngAfterViewChecked(): void {
+    try {
+      if (this.fragment) {
+        document.querySelector('#' + this.fragment).scrollIntoView();
+      }
+    } catch (e) {}
+  }
+
+  removeSpaces(id: string) {
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: It's not a snake case, so to eliminate any misunderstanding changed name on removeSpacesAddUnderscores

File: src/app/ruleset-sample/ruleset-sample.component.ts [hunk 10]
```diff
@@ -46,4 +46,36 @@ export class RulesetSampleComponent implements OnInit, AfterViewChecked {
...
+        return '"' + el + '"';
+      }).join(', ');
+    }
+  }
+
+  getLink(link: string) {
+    return link.split('/').slice(-1)[0];
+  }
+
+  getCondition(condition: any) {
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: Agree, removed getCondition from analysis and experiment

File: src/app/ruleset-sample/ruleset-sample.component.ts [hunk 11]
```diff
@@ -11,6 +11,9 @@ import {ActivatedRoute} from '@angular/router';
 export class RulesetSampleComponent implements OnInit, AfterViewChecked {
   error: string;
   data: any;
+  all_data: any;
+  mandatory_data: any;
+  clicked = false;
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: ok, changed on mandatory_only

File: src/app/ruleset-sample/ruleset-sample.component.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (wizardfan) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: Snake case (or snake_case) is the practice of writing compound words or phrases in which the elements are separated with one underscore character (_) and no spaces, with each element's initial letter usually lowercased within the compound and the first letter either upper- or lowercase—as in "foo_bar" and "Hello_world".
After re-reading the definition and the algorithm, agree that it is not a snake case, as 'two  space" will become "two__space" and snake case will only have one underscore. However does it need two underscores here?

File: src/app/ruleset-experiment/ruleset-experiment.component.html (hunk 1 above)
Comment (Alexey-ebi) [reply src/app/ruleset-experiment/ruleset-experiment.component.html]: Done

File: src/app/ruleset-sample/ruleset-sample.component.html [hunk 12]
```diff
@@ -42,7 +42,7 @@ <h2>FAANG Rule sets
           Show only mandatory fields
         </label>
         <button type="button" class="btn btn-success">
-          Download current template
+          <a href="https://www.ebi.ac.uk/seqdb/confluence/download/attachments/36769258/faang_sample_metadata_empty_template_20181212.xlsx?version=1&modificationDate=1544608044067&api=v2">Download template</a>
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.html]: Moved these links to constants

File: src/app/ruleset-sample/ruleset-sample.component.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: ok, agree, changed on snake case

File: src/app/shared/header/header.component.html [hunk 13]
```diff
@@ -28,7 +28,7 @@
         <a class="nav-link" routerLink="/summary/organisms">Summary</a>
       </li>
       <li class="nav-item" [ngClass]="isActiveRuleset()">
-        <a class="nav-link" routerLink="/ruleset/samples">Rule sets</a>
+        <a class="nav-link" routerLink="/ruleset/samples" fragment="standard">Rule sets</a>
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/shared/header/header.component.html]: Discussed this, have to use fragment

File: src/app/ruleset-analysis/ruleset-analysis.component.html [hunk 14]
```diff
@@ -1,7 +1,7 @@
 <div class="container">
   <app-header></app-header>
   <h2>FAANG Rule sets
-    <button type="button" class="btn btn-info" aria-label="Left Align" [routerLink]="['/ruleset', 'samples']">
+    <button type="button" class="btn btn-info" aria-label="Left Align" [routerLink]="['/ruleset', 'samples']" fragment="standard">
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-analysis/ruleset-analysis.component.html]: Discussed this, have to use fragment

File: src/app/ruleset-sample/ruleset-sample.component.html [hunk 15]
```diff
@@ -37,7 +37,7 @@ <h2>FAANG Rule sets
     <div class="col-md-2"><h3>Rule groups</h3></div>
     <div class="col-md-10">
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.html]: renamed function to mandatoryOnlyToggle

File: src/app/ruleset-sample/ruleset-sample.component.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (wizardfan) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: some misunderstanding here
I mean to change clicked to mandatory_only
the original mandatory_data is absolutely fine

[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply src/app/ruleset-sample/ruleset-sample.component.ts]: Changed mandatory_only to mandatory_data back
Changed clicked to mandatory_only

**Comment by wizardfan** [comment]
Given the planned branch(es) which will 
1. refactor summary, protocol
2. make navbar dropdown to group menu items (currently the list is too long)
3. extract common parts of rulesets/summary/protocol into common component
4. add unit tests
5. separate common functions from constants.ts
this PR is ready to go.

**Pull request merged**